### PR TITLE
Merging of regular and LTS extreme/periodic res1d files

### PR DESCRIPTION
**Pull request by gedaskir**
This PR adds support for merging regular and LTS extreme/periodic res1d files.

### CODE REVIEW

File: util/DHI.Mike1D.MikeIO/LTSResultEvents.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    /// </summary>
+    public void SortOnValue()
+    {
+      Sort(CompareValue);
+    }
+
+    /// <summary>
+    /// Sort on first value, and if the same, then on time.
+    /// </summary>
+    public int CompareValue(LTSResultEvent e1, LTSResultEvent e2)
```
Comment (ecomodeller) [inline util/DHI.Mike1D.MikeIO/LTSResultEvents.cs:22]: Why not use [`IComparable<LTSResultEvent>`](https://learn.microsoft.com/en-us/dotnet/api/system.icomparable-1?view=net-8.0)?

Comment (gedaskir) [reply util/DHI.Mike1D.MikeIO/LTSResultEvents.cs:22]: In Mike1D there is already LTSEvents class in DHI.Mike1D.Engine, which has this type of structure. Now I don't want to change it there. Additionally, there are two ways of sorting which need to be performed. SortOnValue and SortOnTimePeriod. Of course one could say, that it is possible to have LTSResult**s**Extreme and LTSResult**s**Periodic, which implement IComparable<LTSResultEvent> and then has the Sort method. This holds only if LTSResultExtreme does not need sorting on time in DHI.Mike1D.Engine and I think that is the case.

File: tests/test_res1d_lts_events.py [hunk 2]
```diff
@@ -216,3 +219,44 @@ def test_res1d_filter_readall(test_file_path, helpers):
...
+    b4_1491l1_event2 = df_reach.iloc[1].iloc[0]
+    assert b4_1491l1_event1 == b4_1491l1_event2
+    assert pytest.approx(np.round(b4_1491l1_event1, 3)) == 0.151
+
+    df_reach_time = res1d.reaches.B4_1491l1.m_216.DischargeMaximumTime.read()
+    b4_1491l1_time1 = df_reach_time.iloc[0]
+    b4_1491l1_time2 = df_reach_time.iloc[1]
+    assert (b4_1491l1_time1 == b4_1491l1_time2).values[0]
+
+    # Validate all merged events. Every event now needs to appear twice.
```
[the remark this answers is not in the document]
Comment (gedaskir) [reply tests/test_res1d_lts_events.py:257]: The double counting is not a use-case, but a good test, to see that all network locations considered.

**Comment by gedaskir** [comment]
It is a good idea to explain in more detail how the merging happens. 

>     * Is it required that the res1d files have the same network structure? Can we merge different sub-models (with different network structures) into one?
Yes, same network structure is required. Currently, we cannot merge different sub-models.
>     * For standard res1d results, how are overlapping time step values handled?
Currently, for merging standard res1d files I took the typical example of merging files where the simulation end time of one file matches the simulation start of another. I think we should improve to allow merging when the files are apart in time.
>     * For LTS events, how are overlapping events handled? Similarly, how are duplicate events handled?
As you saw from unit tests, currently overlapping in time is allowed and is used as a test to see that all network locations were considered. Eventually, I think as a default option we should throw an exception if files overlap in time.
>     * For LTS chronological, how are overlapping periods handled? Similarly how are duplicate periods handled?
Same as above.

File: tests/test_res1d_lts_chronological.py [hunk 3]
```diff
@@ -254,3 +254,28 @@ def test_global_data_attributes(test_file):
...
+    file_names = [test_file_path, test_file_path]
+    merged_file_name = test_file_path.replace(".res1d", ".merged.res1d")
+    Res1D.merge(file_names, merged_file_name)
+
+    # Read the merged file
+    res1d = Res1D(merged_file_name)
+
+    # Test one reach location for particular values
+    df_reach = res1d.reaches.B4_1320l1.m_101_251.DischargeIntegratedMonthly.read()
+    assert pytest.approx(np.round(df_reach.max(), 3)) == 2 * 1215.915
```
[the remark this answers is not in the document]
Comment (gedaskir) [reply tests/test_res1d_lts_chronological.py:270]: Yes, that is correct, we do not filter out overlapping periods and we double count.

File: tests/test_res1d_catchments.py [hunk 4]
```diff
@@ -221,3 +229,18 @@ def test_catchment_static_attributes(res1d_catchments):
...
+
+def test_res1d_merging(test_file_path_for_merging):
+    file_a = test_file_path_for_merging.replace(".res1d", "_a.res1d")
+    file_b = test_file_path_for_merging.replace(".res1d", "_b.res1d")
+    file_c = test_file_path_for_merging.replace(".res1d", "_c.res1d")
+
+    Res1D.merge([file_a, file_b], file_c)
+
+    df_a = Res1D(file_a).read()
+    df_b = Res1D(file_b).read().tail(-1)
```
[the remark this answers is not in the document]
Comment (gedaskir) [reply tests/test_res1d_catchments.py:242]: Currently they have to. However, we do not validate that. Eventually, this restriction should be lifted.

File: mikeio1d/res1d.py [hunk 5]
```diff
@@ -440,3 +441,18 @@ def to_txt(
     ):
         """Extract to txt file."""
         self.extract(file_path, queries, time_step_skipping_number, ExtractorOutputFileType.TXT)
+
+    @staticmethod
+    def merge(file_names, merged_file_name):
```
[the remark this answers is not in the document]
Comment (gedaskir) [reply mikeio1d/res1d.py:446]: Good idea :-)

**Pull request merged**